src/js/smtpserver.c

```diff
@@ -1,19 +1,154 @@
 #include "smtpserver.h"
 
+jsval create_response(JSContext *cx, int code, const char* message, int disconnect) { 
+	jsval rcode;
+	jsval rmessage;
+	JSObject *obj;
+	jsval rdisconnect; 	// better name? 
+				
+	rcode = INT_TO_JSVAL(code);
+	rdisconnect = disconnect == TRUE ? JSVAL_TRUE : JSVAL_FALSE;
+	
+	obj = JS_NewObject(cx, NULL, NULL, NULL);
+	
+	if (message != NULL) {
+		rmessage = STRING_TO_JSVAL(JS_InternString(cx, message));
+	} else {
+		// TODO
+		// define message property with default value for current code
+		rmessage = STRING_TO_JSVAL(JS_InternString(cx, "default err message"));
+	}
+	
+	JS_DefineProperty(cx, obj, "code", rcode, NULL, NULL, JSPROP_ENUMERATE | JSPROP_READONLY | JSPROP_PERMANENT);
+	
+	JS_DefineProperty(cx, obj, "message", rmessage, NULL, NULL, JSPROP_ENUMERATE | JSPROP_READONLY | JSPROP_PERMANENT);
+	
+	JS_DefineProperty(cx, obj, "disconnect", rdisconnect, NULL, NULL, JSPROP_ENUMERATE | JSPROP_READONLY | JSPROP_PERMANENT);
+	
+	return OBJECT_TO_JSVAL(obj);
+}
+
+JSBool smtpInit(JSContext *cx, unsigned argc, jsval *vp) {
```

Comment:
Since the stub functions are almost identical, it's better to use a preprocessor macro that takes the SMTP command name as argument. You can use preprocessor stringification and implicit string constant concatenation (by the compiler) to form the 3rd argument to `create_response()`. You can use preprocessor concatenation to form the function name. Example:
```c
#define DEFINE_HANDLER_STUB(name) \
    JSBool smtp##name(... \
    jsval rval = create_response(cx, 250, "c" #name", FALSE); \
   etc

DEFINE_HANDLER_STUB(Init)
DEFINE_HANDLER_STUB(Auth)
etc
```